src/HeapJobOrderer.ts

```diff
@@ -0,0 +1,156 @@
+import { Heap } from 'heap-js'
+import { Job } from './Job'
+import { JobOrderer } from './JobOrderer'
+import assert = require('assert')
+
+/**
+ * Manages a set of jobs that have to be run.
+ */
+export class HeapJobOrderer implements JobOrderer {
+	private sourcesHeap: Heap<Job>
+
+	private nonSources: Set<Job> = new Set()
+
+	private inProgress: Set<Job> = new Set()
+
+	private jobToDependents: Map<Job, Set<Job>> = new Map<Job, Set<Job>>()
+
+	private jobToCompletedPrereqs: Map<Job, Set<Job>> = new Map<Job, Set<Job>>()
+
+	/**
+	 * Create the JobOrderer and inform it of the jobs to manage.
+	 *
+	 * @throws When one of the jobs has a prerequisite that is not one of the jobs, or is itself.
+	 * @param jobs - The jobs to manage. They must have correctly populated prerequisites fields.
+	 */
+	constructor(rootJobs: Job[]) {
+		// Initialize the sources heap with its sorting function.
+		this.sourcesHeap = new Heap((job1, job2) => {
+			const job1Dependents = this.jobToDependents.get(job1)
+			const job2Dependents = this.jobToDependents.get(job2)
+
+			assert(
+				job1Dependents && job2Dependents,
+				'Provided jobs are missing from map.',
+			)
+
+			return job2Dependents.size - job1Dependents.size
+		})
+
+		// Add all of the root jobs to dependents map.
+		for (const rootJob of rootJobs) {
+			this.jobToDependents.set(rootJob, new Set())
+		}
+
+		const queue: Job[] = rootJobs
+		const seenSources: Set<Job> = new Set() // We use this because checking whether a heap has an element is O(n).
+
+		// BFS traverse dependency graph, starting with root nodes (the ultimate jobs).
+		while (queue.length) {
+			const job = queue.pop()
```

Comment:
This will modify the `rootJobs` array passed in as an argument; you probably want to make a copy first.